22.0.0.13-beta draft (#2777)
OpenLiberty/blogs: posts/2022-12-06-22.0.0.13-beta.adoc

# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/mbroz2
author_github: https://github.com/mbroz2
seo-title: Jakarta EE 10 Platform and MicroProfile 6 - OpenLiberty.io
seo-description: Open Liberty 22.0.0.13-beta makes available Jakarta EE Platform 10.0, and with it Web Profile 10.0, Jakarta Authentication 3.0, Jakarta Authorization 2.1, Jakarta Security 3.0, and Jakarta Faces 4.0.  MicroProfile 6 is also included in this beta release, as well as the ability to configure FFDC maximum age.
blog_description: Open Liberty 22.0.0.13-beta makes available Jakarta EE Platform 10.0, and with it Web Profile 10.0, Jakarta Authentication 3.0, Jakarta Authorization 2.1, Jakarta Security 3.0, and Jakarta Faces 4.0.  MicroProfile 6 is also included in this beta release, as well as the ability to configure FFDC maximum age.
open-graph-image: https://openliberty.io/img/twitter_card.jpg
open-graph-image-alt: Open Liberty Logo
---
= Jakarta EE 10 Platform and MicroProfile 6
Michal Broz <https://github.com/mbroz2>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

> With Open Liberty 22.0.0.13-beta we're excited to announce many new features and functionality, begging with Jakarta EE Web Profile 10.0 and Jakarta EE Platform 10.0! In order to make that possible, this release also includes the last remaining Jakarta EE 10 features that had not been previously part of a beta: Jakarta Authentication 3.0, Jakarta Authorization 2.1, Jakarta Security 3.0, and Jakarta Faces 4.0.

Reviewer: ```suggestion
With Open Liberty 22.0.0.13-beta we're excited to announce many new features and functions, beginning with Jakarta EE Web Profile 10.0 and Jakarta EE Platform 10.0! To make that possible, this release also includes the last remaining Jakarta EE 10 features that had not previously been part of a beta: Jakarta Authentication 3.0, Jakarta Authorization 2.1, Jakarta Security 3.0, and Jakarta Faces 4.0.

```

> Similarly to Jakarta EE 10 features, we've also been making available various MicroProfile 6 features in the recent beta releases, and 22.0.0.13-beta marks the first release which makes available all of MicroProfile 6!  Also provided in this beta release is the ability to configure maximum FFDC age before it is deleted.

Reviewer: ```suggestion
Similar to Jakarta EE 10 features, we've also been releasing various MicroProfile 6 features in the recent beta releases, and the 22.0.0.13-beta marks the first release that includes all of MicroProfile 6!  Also provided in this beta release is the ability to automatically purges FFDC log files after they reach a configured age.

```

* <<jakarta10, Jakarta EE 10 Platform and Web Profile>> 
** <<authentication, Application Authentication 3.0 (Jakarta Authentication 3.0)>>
** <<authorization, Application Authorization 2.1 (Jakarta Authorization 2.1)>>
** <<security, Application Security 5.0 (Jakarta Security 3.0)>>
** <<faces, Jakarta Faces 4.0>>
* <<mp6, MicroProfile 6.0>>
* <<maxFfdcAge, Configurable maximum FFDC age>>

See also link:{url-prefix}/blog/?search=beta&key=tag[previous Open Liberty beta blog posts].

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // // 
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/23358
// Contact/Reviewer: ReeceNana,jhanders34
// // // // // // // // 
[#jakarta10]
== Jakarta EE 10 Platform and Web Profile

In the Open Liberty 22.0.0.9-beta, we made available Jakarta EE 10 Core Profile and ever since we've been hard at work releasing more Jakarta EE 10 features with each subsequent beta.  We're proud to announce that the 22.0.0.13-beta is the culmination of that effort and makes available Jakarta EE 10.0 Application Client, Jakarta EE Web Profile 10.0, and Jakarta EE Platform 10.0.

> Liberty provides convenience features for running all of the component specifications that are contained in the Jakarta EE Web Profile and Jakarta EE Platform.  `webProfile-10.0` and `jakartaee-10.0` Liberty features provide the Jakarta EE Web Profile and Jakarta EE Platform version 10 Liberty features respectively.  To use Jakarta EE 10 features in the application client, the `jakartaeeClient-10.0` Liberty feature is used.

Reviewer: ```suggestion
Liberty provides convenience features for running all of the component specifications that are contained in the Jakarta EE Web Profile and Jakarta EE Platform.  `webProfile-10.0` and `jakartaee-10.0` Liberty features provide the Jakarta EE Web Profile and Jakarta EE Platform version 10 Liberty features respectively.  For Jakarta EE 10 features in the application client, use the `jakartaeeClient-10.0` Liberty feature.

```

> Using these convenience features enables developers to rapidly develop applications using all of the APIs contained in the Web Profile and Platform specifications.  Unlike the `webProfile-9.1`Liberty feature, the `webProfile-10` Liberty feature does not enable the XML Binding feature since XML Binding is not a Web Profile component specification.  If your application relies on XML Binding APIs, you will need to explicitly enable the `xmlBinding-4.0` Liberty feature when using the `webProfile-10.0` Liberty feature.

Reviewer: ```suggestion
Using these convenience features enables developers to rapidly develop applications using all of the APIs contained in the Web Profile and Platform specifications.  Unlike the `webProfile-9.1` Liberty feature, the `webProfile-10` Liberty feature does not enable the XML Binding feature, because XML Binding is not a Web Profile component specification.  If your application relies on XML Binding APIs, you must explicitly enable the `xmlBinding-4.0` Liberty feature when using the `webProfile-10.0` Liberty feature.

```

To enable the Jakarta EE 10 beta features in your application’s `server.xml`:
[source,xml]
----
  <featureManager>
    <feature>jakartaee-10.0</feature>
  </featureManager>
----
Or you can add the Web Profile convenience feature to enable all of the Jakarta EE 10 Web Profile beta features at once:
[source,xml]
----
  <featureManager>
    <feature>webProfile-10.0</feature>
  </featureManager>
----
To run Jakarta EE 10 features on the Application Client Container in your application's `client.xml`:
[source,xml]
----
 <featureManager>
       <feature>jakartaeeClient-10.0</feature>
 </featureManager>
----
For more information see the https://jakarta.ee/specifications/platform/10/[Jakarta EE Platform 10 Specification] and the https://jakarta.ee/specifications/webprofile/10/[Jakarta EE Web Profile 10 Specification].
    
    
// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC> 

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // // 
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/23356
// Contact/Reviewer: ReeceNana,arkarkala
// // // // // // // // 
[#authentication]
== Application Authentication 3.0 (Jakarta Authentication 3.0)

> Jakarta Authentication defines a general low-level SPI for authentication mechanisms, which are controllers that interact with a caller and a container's environment to obtain the caller's credentials, validate these, and pass an authenticated identity (such as name and groups) to the container.

Reviewer: ```suggestion
Jakarta Authentication defines a general low-level SPI for authentication mechanisms, which are controllers that interact with a caller and a container's environment to obtain and validate the caller's credentials and pass an authenticated identity (such as name and groups) to the container.

```

> Multiple enhancements are made available in Jakarta Authentication 3.0, including providing methods to add and remove single server auth module, and the addition of default methods to `ServerAuth` and `ClientAuth` modules.

Reviewer: ```suggestion
Multiple enhancements are made available in Jakarta Authentication 3.0, including methods to add and remove single server auth module, and the addition of default methods to `ServerAuth` and `ClientAuth` modules.

```

   
[source,xml]
----
    <featureManager>
        <feature>appAuthorization-2.1</feature>
    </featureManager>
----
   
For more information see the https://github.com/jakartaee/specifications/blob/master/authorization/2.1/_index.md[Jakarta Authorization 2.1] specification.    

// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC> 

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // // 
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/23315
// Contact/Reviewer: teddyjtorres,ReeceNana
// // // // // // // // 
[#security]
== Application Security 5.0 (Jakarta Security 3.0)

Application Security 5.0 enables support for securing the server runtime environment and applications using Jakarta Security 3.0. The Jakarta Security 3.0 specification allows using built-in authentication mechanisms and identity stores as well as providing your own custom implementations for authenticating users to protected resources. The specification introduces a new built-in authentication mechanism based on the Authorization Code Flow of the OpenID Connect Core 1.0 specification.

> Jakarta Security 3.0 allows using the new `@OpenIdAuthenticationMechanismDefinition` annotation to configure a built-in OpenID Connect authentication mechanism to function as an OpenID Connect client, or Relying Party (RP), so that you can use an OpenID Connect Provider (OP) as a single-sign-on identity provider. It also introduces the `OpenIdContext` bean and various APIs for handling the access token, identity token, and refresh token.

Reviewer: ```suggestion
Jakarta Security 3.0 allows using the new `@OpenIdAuthenticationMechanismDefinition` annotation to configure a built-in OpenID Connect authentication mechanism to function as an OpenID Connect client, or Relying Party (RP), so that you can use an OpenID Connect Provider (OP) as a single sign-on identity provider. It also introduces the `OpenIdContext` bean and various APIs for handling the access token, identity token, and refresh token.

```

   
You can use the `@OpenIdAuthenticationMechanismDefinition` annotation in a Web or JAX-RS application to enable an OpenID Connect client for that application. The following example shows a servlet with a minimal `@OpenIdAuthenticationMechanismDefinition` annotation when the OP is a Liberty OP:

[source,java]
----
@WebServlet("/OidcAnnotatedServlet")
@OpenIdAuthenticationMechanismDefinition(
                providerURI = "https://localhost:8920/oidc/endpoint/OP",
                clientId = "clientId",
                clientSecret = "clientSecret",
                redirectURI = "https://localhost:8940/MyApp/Callback",
                claimsDefinition = @ClaimsDefinition(callerNameClaim = "sub", callerGroupsClaim = "groupIds"))
@DeclareRoles("all")
@ServletSecurity(@HttpConstraint(rolesAllowed = "all"))
public class OidcAnnotatedServlet extends HttpServlet {
…
}
----

You can enable the Application Security 5.0 (Jakarta Security 3.0) feature by using the `appSecurity-5.0` feature in the server.xml file:
   
[source,xml]
----
    <featureManager>
        <feature>appSecurity-5.0</feature>
    </featureManager>
----

> For more information about the `@OpenIdAuthenticationMechanismDefinition` annotation and its attributes refer to the section https://jakarta.ee/specifications/security/3.0/jakarta-security-spec-3.0.html#openid-connect-annotation[2.4.4. OpenID Connect Annotation] of the Jakarta Security 3.0 specification. Please refer to https://javadoc.io/doc/jakarta.security.enterprise/jakarta.security.enterprise-api/latest/jakarta.security/module-summary.html[Jakarta Security Module] for the documentation of the Jakarta Security 3.0 APIs. 

Reviewer: ```suggestion
For more information about the `@OpenIdAuthenticationMechanismDefinition` annotation and its attributes, refer to the https://jakarta.ee/specifications/security/3.0/jakarta-security-spec-3.0.html#openid-connect-annotation[2.4.4. OpenID Connect Annotation] section of the Jakarta Security 3.0 specification. Please refer to https://javadoc.io/doc/jakarta.security.enterprise/jakarta.security.enterprise-api/latest/jakarta.security/module-summary.html[Jakarta Security Module] for the documentation of the Jakarta Security 3.0 APIs. 

```

  
// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC> 

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // // 
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/23224
// Contact/Reviewer: volosied,pnicolucci,ReeceNana
// // // // // // // // 
[#faces]
== Jakarta Faces 4.0

Jakarta Faces is a Model-View-Controller (MVC) framework for building web applications. It offers many convenient features, such as state management and input validation.

> Numerous changes have occurred since the last Jakarta Faces release. Notably, Pages (formerly JSP) support has been removed. Managed Beans support has also been removed, and developers should use CDI beans from this version forward. The namespaces also have been updated from `http://xmlns.jcp.org/jsf/*` to `jakarta.faces.*`. Otherwise, new attributes have been added to existing tags: `multiple` and `accept` for `h:inputFile` and `onerror` for `f:websocket`. Developers may now create facelets programmatically rather than building them via XML. Extensionless mapping has been implemented, so URLs no longer need to include `/faces` or `.xhtml` mappings. Two new tags have been added: `f:selectItemGroups` and `f:selectItemGroup` to create a nested collections (i.e., optgroup containing option elements).  These are just some significant changes, but there have been many more.

Reviewer: ```suggestion
Numerous changes have occurred since the last Jakarta Faces release. Notably, Pages (formerly JSP) support has been removed. Managed Beans support has also been removed, and developers should use CDI beans from this version forward. The namespaces also have been updated from \`http://xmlns.jcp.org/jsf/*` to \`jakarta.faces.*`. Otherwise, new attributes have been added to existing tags: `multiple` and `accept` for `h:inputFile` and `onerror` for `f:websocket`. Developers can now create facelets programmatically rather than building them via XML. Extensionless mapping is implemented, so URLs no longer need to include `/faces` or `.xhtml` mappings. Two new tags have been added to create a nested collections (i.e., optgroup containing option elements): `f:selectItemGroups` and `f:selectItemGroup`.  These are just some significant changes, but there are many more.

```

Reviewer: need to escape those *  in the namespaces as they are not rendering in preview (and probably not on the site either. I think a forward slash \ does it but not sure

Reviewer 2: I'll make sure keep an eye out on the draft site to see how it comes out with the proposed changed

> The `faces-4.0` feature uses MyFaces, but if a developer wishes to use an alternative Faces implementation, such as Mojarra 4.0, that can be accomplished via facesContainer-4.0. The Faces API and implementation jars just need to be included in the application `WEB-INF/lib` folder. 

Reviewer: ```suggestion
The `faces-4.0` feature uses MyFaces, but if you want to use an alternative Faces implementation, such as Mojarra 4.0, you can use the `facesContainer-4.0` feature instead. The Faces API and implementation jars just need to be included in the application `WEB-INF/lib` folder. 

```

The following specifications in MicroProfile 6.0 release are either new or have some major or minor update when compared to MicroProfile 5.0:

- MicroProfile Telemetry 1.0 (new spec)
- MicroProfile Metrics 5.0 (major update)
- MicroProfile OpenAPI 3.1 (minor update)
- MicroProfile JWT 2.1 (minor update)

MicroProfile 6.0 has the following backward incompatible changes compared to MicroProfile 5.0:

- Incompatible changes from MicroProfile Metrics
- MicroProfile OpenTracing replaced by MicroProfile Telemetry
- Includes Jakarta EE 10 Core Profile instead of the five specifications (CDI, Jakarta Restful Web Services, JSON-B, JSON-P and Annotations) in Jakarta EE 9.1

It's also worth noting that MicroProfile Metrics 5.0 has https://download.eclipse.org/microprofile/microprofile-metrics-5.0.0-RC4/microprofile-metrics-spec-5.0.0-RC4.html#_incompatible_changes[incompatible changes] compared to MicroProfile Metrics 4.0.

You can enable the MicroProfile 6 feature by using the `microProfile-6.0` feature in the server.xml file:

[source,xml]
----
<featureManager>
    <feature>microProfile-6.0</feature>
</featureManager>
----
    

> To find out more take a look at the https://github.com/eclipse/microprofile/releases/tag/6.0-RC3[MicroProfile 6.0 Release].

Reviewer: ```suggestion
To find out more, take a look at the https://github.com/eclipse/microprofile/releases/tag/6.0-RC3[MicroProfile 6.0 Release].

```

    
// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC> 

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // // 
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/23235
// Contact/Reviewer: ReeceNana,tonyreigns
// // // // // // // // 
[#maxFfdcAge]
== Configurable maximum FFDC age
   

> Open Liberty provides First Failure Data Capture(FFDC) capability which instantly collects information about events and conditions that might lead up to a failure.  In certain scenarios, the number of files in the FFDC directory can grow to a very large amount. Currently, Open Liberty will automatically purge FFDC files in excess of 500 and this value was not configurable. This beta release introduces new functionality which automatically purges FFDC log files after reaching a configured age. This is done through a new configuration attribute, `maxFfdcAge`.

Reviewer: ```suggestion
Open Liberty provides First Failure Data Capture(FFDC) capability which instantly collects information about events and conditions that might lead up to a failure.  In certain scenarios, the number of files in the FFDC directory can grow to a very large amount. Currently, Open Liberty automatically purges FFDC files in excess of 500 and this value is not configurable. This beta release introduces new functionality which automatically purges FFDC log files after they reach a configured age. This is done through a new configuration attribute, `maxFfdcAge`.

```
